### PR TITLE
re #30: Switch to xeus-python kernel

### PR DESCRIPTION
**Pull request by lassoan**
Replace xSlicerInterpreter with xeus-python's interpreter. xeus-python implements many more features and simplifies our code. It works really well - widgets can now be used in Slicer notebooks, syntax highlighting, documentation, stream redirection, error handling, etc. all work very nicely!

To do before replacing the old implementation:
- [x] Figure out a mechanism to display output in Slicer's Python console when commands are entered in Slicer's Python console. Maybe always sending output to both the notebook and Slicer's Python console would be acceptable. (requires merge of https://github.com/jupyter-xeus/xeus-python/pull/269)
- [x] Implement views display (3D, slice, etc.) using standard display() method
- [x] Implement node display, for example display a table node as a table in the notebook
- [x] Fix inspection of Python-wrapped VTK classes. Currently jedi cannot extract documentation from Python-wrapped VTK classes (this may be fixed in more recent VTK versions). In the short term we should detect if jedi could not return documentation and send back the help text of the method or class (as we did in the original xSlicerInterpreter implementation).
- [x] Fix display visualization in Slicer's Python console (display results are now only displayed in the notebook - such as `slicer.mrmlScene`)
- [x] Review usage of global interpreter lock, test robustness in various build modes and platforms. There have been issues with hanging/crashing at startup in debug mode, but maybe that was due to a messed up Slicer build.
- [ ] Review and test build system and installation

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
This is great! Really happy to see this go through! If anything comes up in core xeus-python, please do not hesitate to report back!

**Comment by SylvainCorlay** [comment]
Note: it also works with custom widgets such as ITK... I don't know if this makes sense together with Slicer, but there might be some really nice demos ahead.

File: JupyterKernel/qSlicerJupyterKernelModule.cxx [hunk 1]
```diff
@@ -290,6 +239,171 @@ def inspect(code, cursor_pos, detail_level):
...
+    bArray = qt.QByteArray()
+    buffer = qt.QBuffer(bArray)
+    buffer.open(qt.QIODevice.WriteOnly)
+    screenshot.save(buffer, "PNG")
+    dataValue = bArray.toBase64().data().decode()
+    dataType = "image/png"
+    return { dataType: dataValue }
+
+
+ # empty object to store Slicer notebook helper functions
```
Comment (jcfr) [inline JupyterKernel/qSlicerJupyterKernelModule.cxx:386]: indent

File: JupyterKernel/qSlicerJupyterKernelModule.cxx [hunk 2]
```diff
@@ -290,6 +239,171 @@ def inspect(code, cursor_pos, detail_level):
...
+slicer.nb.displayViews = displayViews
+slicer.nb.displayModel = displayModel
+slicer.nb.displayTable = displayTable
+
+# Set up display hook to display execution result
+# as raw text in Slicer's Python console and as
+# rich output in Jupyter notebook.
+
+def slicerDisplayHook(value):
+  # show raw output in Slicer's Python console
```
Comment (jcfr) [inline JupyterKernel/qSlicerJupyterKernelModule.cxx:398]:
```
  """Show raw output in Slicer's Python console
  """
```

**Comment by jcfr** [comment]
> Note: it also works with custom widgets such as ITK... I don't know if this makes sense together with Slicer, but there might be some really nice demos ahead.

@lassoan I am thinking that we could make use of [itkwidgets](https://github.com/InsightSoftwareConsortium/itkwidgets#readme) to interactively visualize data in the notebook. For example, model could be visualized interactively instead of being rendered as a static image.

I  am even thinking we could have some "special" view associated with node like `vtkMRMLRedSliceNode`, `vtkMRMLViewNode`, ...  that could leverage html  vnc. 

These are just idea. No blocker.

@thewtex What do you think ?

**Comment by jcfr** [comment]
It is very exciting ! Thanks @lassoan @ihnorton @SylvainCorlay @martinRenou for your help with this :pray: 

We got a long way since SciPy sprint in 2014 when this crazy idea started. Also thanks to @msarahan who was crazy enough to joined the sprint team, listen and support it (for reference .... see https://github.com/commontk/QEmbedIPython)

**Comment by lassoan** [comment]
We can now use Slicer as a remote renderer, with interactive widgets, which may work well even for huge data sets or complex rendering scenes:

image

itkwidgets will be better for quick interactive rendering of small datasets and I was really excited to try them. Unfortunately, I could not make them work at all - view widgets did not appear. It does not seem to be a problem in Slicer's Python environment only. itkwidgets' own binder demos did not work for me either: I've checked out 20-25 examples and only one of them worked. @thewtex are you aware of these issues? Do you have any suggestions how to solve them?

**Comment by lassoan** [comment]
All issues have been fixed and branch is merged.

@jcfr I have only tested this on Windows and there is a build error (python packages not found) but when I run the build again it succeeds. It might cause problems when factory tries to build it. It would be awesome if you could test on Linux and have a look at this build error. I would really like to get this into hands of users soon!

**Comment by jcfr** [comment]
Building on windows using VS2015 fails while building xeus-python with the following error:

```
2>  D:\D\P\S-0-E-b\SlicerJupyter-build\xeus-python\src\xptvsd_client.cpp(274): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
2>D:\D\P\S-0-E-b\SlicerJupyter-build\xeus-python\src\xptvsd_client.cpp(312): error C2664: 'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer> nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>::parse(nlohmann::detail::input_adapter &&,const std::function<bool (int,nlohmann::detail::parser<nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>>::parse_event_t,BasicJsonType &)>,const bool)': cannot convert argument 1 from 'const std::string' to 'nlohmann::detail::input_adapter &&'
2>          with
2>          [
2>              BasicJsonType=nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer>
2>          ]
2>  D:\D\P\S-0-E-b\SlicerJupyter-build\xeus-python\src\xptvsd_client.cpp(312): note: Reason: cannot convert from 'const std::string' to 'nlohmann::detail::input_adapter'
2>  D:\D\P\S-0-E-b\SlicerJupyter-build\xeus-python\src\xptvsd_client.cpp(312): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called

```

**Comment by SylvainCorlay** [comment]
> Building on windows using VS2015 fails while building xeus-python with the following error:

cc @martinRenou this rings a bell.

@jcfr maybe this is resolved with a more recent version of nlohmann_json.

**Comment by jcfr** [comment]
I will report back with more details, there is something else going on with the build.

**Comment by jcfr** [comment]
Linux and macOS build as expected, let's wait tomorrow results for windows. 

A clean build indicates that the windows error I reported above seems to be irrelevant.

**Comment by SylvainCorlay** [comment]
@lassoan @jcfr SlicerJupyter may be a really great subject for a guest post on Jupyter's blog!

It can also be an excellent example on how to integrate with Jupyter for other desktop applications that embed a Python interpreter (FreeCAD, Blender, etc).

**Comment by jcfr** [comment]
A guess post would be great, [here is a google doc](https://docs.google.com/document/d/1oSUw_0LOoQSTzFxIDI9KpI2HzzhUhijek4WQYKtIh_0/edit?usp=sharing) to start collaborating text.

**Comment by SylvainCorlay** [comment]
I am happy to help review the post on the xeus aspects - but there are probably many other aspects to it!

**Comment by jcfr** [comment]
As a side note, building on what we did with Slicer, the paraview team has been working on a similar plugin. See https://gitlab.kitware.com/paraview/paraviewjupyterkernel/-/tree/main/ :tada:

**Comment by SylvainCorlay** [comment]
> As a side note, building on what we did with Slicer, @nvuaille has been working on a similar plugin. See https://gitlab.kitware.com/paraview/paraviewjupyterkernel/-/tree/main/ tada

Great! I think this is all coming together really nicely.

**Comment by thewtex** [comment]
> itkwidgets will be better for quick interactive rendering of small datasets and I was really excited to try them. Unfortunately, I could not make them work at all - view widgets did not appear. It does not seem to be a problem in Slicer's Python environment only. itkwidgets' own binder demos did not work for me either: I've checked out 20-25 examples and only one of them worked. @thewtex are you aware of these issues? Do you have any suggestions how to solve them?

@lassoan itkwidgets should work well on modestly sized images, i.e. up to 5GB. Here's a notebook that discusses this:

https://mybinder.org/v2/gh/InsightSoftwareConsortium/itkwidgets/master?urlpath=lab/tree/examples/LargeVolumes.ipynb

We are also working towards support for terabyte-scale images.

Recently, the distributed version was not working due to a packaging issue :package: :exploding_head: , but that was fixed today. Please try:

`pip install --upgrade itkwidgets`

Excited to see what you come with and a demo of how these can be used together.

**Comment by thewtex** [comment]
> but that was fixed today.

While testing locally worked, checking Binder again, there are still issues. I'll address that...

**Comment by nvuaille** [comment]
> > As a side note, building on what we did with Slicer, @nvuaille has been working on a similar plugin. See https://gitlab.kitware.com/paraview/paraviewjupyterkernel/-/tree/main/ tada
> 
> Great! I think this is all coming together really nicely. It may be a good idea to directly use xeus-python in that case as well.

It seems absolutely a good idea, I'm going to have a look at this really soon !

**Comment by SylvainCorlay** [comment]
@nvuaille have you looked into using xeus-python directly rather than xeus in paraview?

**Comment by lassoan** [comment]
@SylvainCorlay I would be happy to write a blog post about this. Xeus-python kernel works really well.

Just one question: It seems then when we want to use widgets then we have to install ipywidgets in the kernel, which pulls in the entire ipython universe. Is this expected?

**Comment by SylvainCorlay** [comment]
> Just one question: It seems then when we want to use widgets then we have to install ipywidgets in the kernel, which pulls in the entire ipython universe. Is this expected?

Unfortunately it is expected.

With ipywidgets 8.0, we cut the dependency of `ipywidgets` to `notebook` (through widgetsnbextension), which should reduce by a lot the number of dependencies.

**Comment by SylvainCorlay** [comment]
> @SylvainCorlay I would be happy to write a blog post about this. Xeus-python kernel works really well.

Great post on the Slicer discourse by the way. I think the screencast would be perfect for a Jupyter story!

**Pull request merged**